kpi/utils/usage_calculator.py

```diff
@@ -0,0 +1,170 @@
+from datetime import timedelta
+from json import loads, dumps
+from typing import Optional
+
+from django.conf import settings
+from django.db.models import Sum, Q
+from django.db.models.functions import Coalesce
+from django.utils import timezone
+
+from kobo.apps.kobo_auth.shortcuts import User
+from kobo.apps.openrosa.apps.logger.models import (
+    DailyXFormSubmissionCounter,
+    XForm,
+)
+from kobo.apps.organizations.models import Organization
+from kobo.apps.organizations.utils import (
+    get_monthly_billing_dates,
+    get_yearly_billing_dates,
+)
+from kobo.apps.stripe.constants import ACTIVE_STRIPE_STATUSES
+from kobo.apps.trackers.models import NLPUsageCounter
+from kpi.utils.cache import CachedClass, cached_class_property
+
+
+class ServiceUsageCalculator(CachedClass):
+    CACHE_TTL = settings.ENDPOINT_CACHE_DURATION
+
+    def __init__(self, user: User, organization: Optional[Organization]):
+        self.user = user
+        self.organization = organization
+
+        self._user_ids = [user.pk]
+        self._user_id_query = self._filter_by_user([user.pk])
+        if organization and settings.STRIPE_ENABLED:
+            # if the user is in an organization and has an enterprise plan, get all org users
+            # we evaluate this queryset instead of using it as a subquery because it's referencing
+            # fields from the auth_user tables on kpi *and* kobocat, making getting results in a
+            # single query not feasible until those tables are combined
+            user_ids = list(
+                User.objects.filter(
+                    organizations_organization__id=organization.id,
+                    organizations_organization__djstripe_customers__subscriptions__status__in=ACTIVE_STRIPE_STATUSES,
+                    organizations_organization__djstripe_customers__subscriptions__items__price__product__metadata__has_key='plan_type',
+                    organizations_organization__djstripe_customers__subscriptions__items__price__product__metadata__plan_type='enterprise',
+                ).values_list('pk', flat=True)[
+                    : settings.ORGANIZATION_USER_LIMIT
+                ]
+            )
+            if user_ids:
+                self._user_ids = user_ids
+                self._user_id_query = self._filter_by_user(user_ids)
+
+        now = timezone.now()
+        self.current_month_start, self.current_month_end = (
+            get_monthly_billing_dates(organization)
+        )
+        self.current_year_start, self.current_year_end = (
+            get_yearly_billing_dates(organization)
+        )
+        self.current_month_filter = Q(
+            date__range=[self.current_month_start, now]
+        )
+        self.current_year_filter = Q(date__range=[self.current_year_start, now])
+        self._setup_cache()
+
+    def _get_cache_hash(self):
+        org_id = self.organization.id if self.organization else None
+        return f'{self.user.id}-{org_id}'
```

Comment:
Could this be wrong if two users belong to the same organization. Correct me if I'm wrong but I think every user in the same org should see the same usage, thus, they should use the same key (and obviously receive the same cache object). 
So, IMHO,  `self.user.id` should not be part of the cache key if organization exists.

Reply:
Yes, you're right. I made it use only the org id if there is any